src/InstallAgent/Device.cs

```diff
@@ -0,0 +1,211 @@
+                    if ((j = Array.FindIndex(hwIDs, hwid => hwIDFound(id, hwid))) != -1)
+                    {
+                        devices[j] = tmpDevInfoData;
+                        tmpDevInfoData = new PInvoke.SetupApi.SP_DEVINFO_DATA();
+                        tmpDevInfoData.cbSize = (uint)Marshal.SizeOf(tmpDevInfoData);
+                        break;
+                    }
+                }
+            }
+
+            return Marshal.GetLastWin32Error();
+        }
+
+        public static void RemoveFromRegistry(string[] hwIDs, bool strictSearch)
+        {
+            IntPtr deviceInfoSet = PInvoke.SetupApi.SetupDiGetClassDevs(
+                IntPtr.Zero,
+                IntPtr.Zero,
+                IntPtr.Zero,
+                (uint)PInvoke.SetupApi.DiGetClassFlags.DIGCF_ALLCLASSES
+            );
+
+            if (deviceInfoSet == INVALID_HANDLE_VALUE)
+            {
+                return;
+            }
+
+            string excMsg = "";
+
+            PInvoke.SetupApi.SP_DEVINFO_DATA[] devices =
+                new PInvoke.SetupApi.SP_DEVINFO_DATA[hwIDs.Length];
+
+            FindInRegistry(ref devices, hwIDs, deviceInfoSet, strictSearch);
+
+            for (int i = 0; i < devices.Length; ++i)
+            {
+                if (devices[i].cbSize == 0)
+                {
+                    continue;
+                }
+
+                Trace.WriteLine("Trying to remove " + hwIDs[i]);
+                PInvoke.SetupApi.REMOVE_PARAMS rparams =
+                    new PInvoke.SetupApi.REMOVE_PARAMS();
+                rparams.cbSize = 8; // Size of cbSide & InstallFunction
+                rparams.InstallFunction = PInvoke.SetupApi.DIF_REMOVE;
+                rparams.HwProfile = 0;
+                rparams.Scope = PInvoke.SetupApi.DI_REMOVE_DEVICE_GLOBAL;
+                GCHandle handle1 = GCHandle.Alloc(rparams);
+
+                if (!PInvoke.SetupApi.SetupDiSetClassInstallParams(
+                        deviceInfoSet,
+                        ref devices[i],
+                        ref rparams,
+                        Marshal.SizeOf(rparams)))
+                {
+                    excMsg = String.Format(
+                        "Unable to set class install params for hardware device: {0}",
+                        hwIDs[i]
+                    );
+
+                    break;
+                }
+
+                if (!PInvoke.SetupApi.SetupDiCallClassInstaller(
+                        PInvoke.SetupApi.DIF_REMOVE,
+                        deviceInfoSet,
+                        ref devices[i]))
+                {
+                    excMsg = String.Format(
+                        "Unable to call class installer for hardware device: {0}",
+                        hwIDs[i]
+                    );
+
+                    break;
+                }
+
+                Trace.WriteLine("Remove should have worked");
+            }
+
+            PInvoke.SetupApi.SetupDiDestroyDeviceInfoList(deviceInfoSet);
```

Comment:
Rather than using break; and falling through, would this be clearer as throwing an exception, then cleaning up inside a finally block?